File: scripts/scrape-ofsted.js
#!/usr/bin/env node

import fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import https from 'node:https';
import cheerio from 'cheerio';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const SEARCH_URL = 'https://reports.ofsted.gov.uk/search?q=&location=&lat=&lon=&radius=&level_1_types=1&level_2_types%5B%5D=3';

function fetch(url) {
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { 'User-Agent': 'Mozilla/5.0 (compatible; OfstedScraper/1.0)' } }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return resolve(fetch(new URL(res.headers.location, url).toString()));
      }
      if (res.statusCode !== 200) {
        reject(new Error(`Request failed with status ${res.statusCode}`));
        return;
      }
      let data = '';
      res.on('data', (chunk) => (data += chunk));
      res.on('end', () => resolve(data));
    }).on('error', reject);
  });
}

function parseResults(html, limit = 10) {
  const $ = cheerio.load(html);
  const items = [];
  $('ul.results-list > li.search-result').each((i, el) => {
    if (items.length >= limit) return false;

    const titleEl = $(el).find('h3.search-result__title a');
    const name = titleEl.text().trim();

    const category = $(el)
      .find('.search-result__provider-info li:contains("Category:") strong')
      .first()
      .text()
      .trim() || null;

    const address = $(el).find('address.search-result__address').text().trim() || null;

    // Rating may not exist
    const rating = $(el)
      .find('.search-result__provider-info li:contains("Rating:") strong')
      .first()
      .text()
      .trim() || null;

    const latestReport = $(el)
      .find('.search-result__provider-info li:contains("Latest report:") time')
      .first()
      .text()
      .trim() || null;

    const urn = $(el)
      .find('.search-result__provider-info li:contains("URN:") strong')
      .first()
      .text()
      .trim() || null;

    const href = titleEl.attr('href');
    const providerUrl = href ? new URL(href, 'https://reports.ofsted.gov.uk').toString() : null;

    items.push({ name, category, address, rating, latestReport, urn, providerUrl });
  });
  return items;
}

async function main() {
  const html = await fetch(SEARCH_URL);
  const results = parseResults(html, 10);
  const outDir = path.resolve(__dirname, '..', 'data');
  await fs.mkdir(outDir, { recursive: true });
  const ts = new Date().toISOString().replace(/[:.]/g, '-');
  const jsonPath = path.join(outDir, `ofsted-fes-first-10-${ts}.json`);
  const csvPath = path.join(outDir, `ofsted-fes-first-10-${ts}.csv`);

  await fs.writeFile(jsonPath, JSON.stringify(results, null, 2), 'utf8');

  const headers = ['Name', 'Category', 'Address', 'Rating', 'Latest report', 'URN', 'Provider URL'];
  const lines = [headers.join(',')];
  for (const r of results) {
    const row = [r.name, r.category, r.address, r.rating, r.latestReport, r.urn, r.providerUrl]
      .map((v) => {
        if (v == null) return '';
        const s = String(v).replace(/"/g, '""');
        return `"${s}"`;
      })
      .join(',');
    lines.push(row);
  }
  await fs.writeFile(csvPath, lines.join('\n'), 'utf8');

  console.log(JSON.stringify({ count: results.length, jsonPath, csvPath }, null, 2));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scrape_ofsted.py
#!/usr/bin/env python3
"""
Scrape Outstanding training providers from Ofsted website
"""

import requests
from bs4 import BeautifulSoup
import csv
import time
from urllib.parse import urljoin

def scrape_ofsted_providers(base_url):
    """
    Scrape training providers with Outstanding rating from Ofsted
    """
    providers = []
    page = 1
    url = base_url
    
    # Set up headers to mimic a browser
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }
    
    while True:
        print(f"Scraping page {page}...")
        
        try:
            response = requests.get(url, headers=headers, timeout=30)
            response.raise_for_status()
            
            soup = BeautifulSoup(response.content, 'html.parser')
            
            # Find all provider results
            results = soup.find_all('li', class_='search-result')
            
            if not results:
                print(f"No results found on page {page}. Stopping.")
                break
            
            print(f"Found {len(results)} results on page {page}")
            
            for result in results:
                provider = {}
                
                # Extract provider name and link
                title_elem = result.find('h3', class_='search-result__title')
                if title_elem:
                    link = title_elem.find('a')
                    if link:
                        provider['Name'] = link.get_text(strip=True)
                        provider['URL'] = urljoin('https://reports.ofsted.gov.uk', link['href'])
                
                # Extract all provider info items (Category, Rating, Latest report, URN)
                info_lists = result.find_all('ul', class_='search-result__provider-info')
                
                for info_list in info_lists:
                    items = info_list.find_all('li')
                    for item in items:
                        text = item.get_text(strip=True)
                        if ':' in text:
                            label, value = text.split(':', 1)
                            label = label.strip()
                            value = value.strip()
                            provider[label] = value
                
                # Extract address
                address_elem = result.find('address', class_='search-result__address')
                if address_elem:
                    provider['Address'] = address_elem.get_text(strip=True)
                
                providers.append(provider)
            
            # Check if there's a next page
            pagination = soup.find('div', class_='pagination')
            if pagination:
                next_button = pagination.find('a', class_='pagination__next')
                if next_button and next_button.get('href'):
                    url = urljoin('https://reports.ofsted.gov.uk', next_button['href'])
                    page += 1
                    time.sleep(1)  # Be polite and wait between requests
                else:
                    print("No more pages. Stopping.")
                    break
            else:
                print("No pagination found. Stopping.")
                break
            
        except Exception as e:
            print(f"Error scraping page {page}: {e}")
            break
    
    return providers

def save_to_csv(providers, filename='ofsted_outstanding_providers.csv'):
    """
    Save providers data to CSV file
    """
    if not providers:
        print("No providers to save")
        return
    
    # Get all unique keys from all providers
    fieldnames = set()
    for provider in providers:
        fieldnames.update(provider.keys())
    
    # Sort fieldnames for consistency, but put important ones first
    priority_fields = ['Name', 'Category', 'Address', 'Rating', 'Latest report', 'URN', 'URL']
    fieldnames = [f for f in priority_fields if f in fieldnames] + sorted(list(fieldnames - set(priority_fields)))
    
    with open(filename, 'w', newline='', encoding='utf-8') as csvfile:
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(providers)
    
    print(f"\nSaved {len(providers)} providers to {filename}")

def main():
    # URL for Outstanding Independent Specialist Colleges
    url = "https://reports.ofsted.gov.uk/search?q=&location=&radius=&level_1_types=1&level_2_types=3&latest_report_date_start=&latest_report_date_end=&rating%5B%5D=1&status%5B%5D=1"
    
    print("Starting Ofsted scraper...")
    print(f"URL: {url}\n")
    
    providers = scrape_ofsted_providers(url)
    
    print(f"\nTotal providers scraped: {len(providers)}")
    
    if providers:
        save_to_csv(providers)
        
        # Print first few providers as sample
        print("\nSample of scraped data:")
        for i, provider in enumerate(providers[:3], 1):
            print(f"\n{i}. {provider.get('Name', 'N/A')}")
            print(f"   Category: {provider.get('Category', 'N/A')}")
            print(f"   Address: {provider.get('Address', 'N/A')}")
            print(f"   Rating: {provider.get('Rating', 'N/A')}")
            print(f"   URN: {provider.get('URN', 'N/A')}")

if __name__ == "__main__":
    main()
